llext: add support for Harvard arch, arcmwdt

laurenmurphyx64 opened the pull request:
Adds support for Harvard architecture and the ARC MetaWare Design Toolchain as tested on nsim/nsim_vpx5. Also adds a script for reordering sections of ET_REL objects (with elftools) when their order is incompatible with the LLEXT loader (regions overlap).

**IMPORTANT NOTE** RE: reordering script: I am not pushing very hard for llext_reorder_sects.py. If people feel like using elftools to rewrite ET_REL objects will compromise their integrity, I can just drop the commit. But why propose it? For most test extensions on most architectures, the compiler generates an object whose sections are in the correct order to be loaded. But some test extensions on certain architectures / toolchains, and I think that more complex extensions our users come up with will also be out of order. Developers will not always be able to correct the order by editing the source file e.g. the header order. Below is a list of test extensions that the compiler generated with sections in an invalid order.

* syscalls for x86 platforms with MMU (can be corrected by changing header order)
* init_fini for nsim/nsim_vpx5 with arcmwdt (can't be corrected)
* All extensions on qemu_malta*
* All extensions on qemu_leon3/leon3*

*We don't support these architectures yet, and currently just build at the moment, but the extensions as-is would fail to load if we tried

A key point to note about the reordering script is that it would require a Kconfig to enable users to use detached sections, because there is no other way for a Python script invoked from CMake to detect a section is detached without it and forego reordering.

**NOTE**: RE: stripping script: To get tests/subsys/llext to work with MWDT on NSIM VPX, I had to shrink the extension sizes from 16k so they would fit into memory. 15k of the extension produced after compiling and stripping was taken up by a massive shstrtab as MWDT's stripac failed to remove the section names of the 700 useless debug sections it stripped. I talked to support from Synopsys and they told me that as there was apparently no way to stop CCAC from emitting these debug sections, I should edit the ELF myself to rebuild the shstrtab. I wrote a script using elftools to do this. 

Much later, I realized I could invoke the Zephyr SDK strip instead of my script. I put a commit at the top of this PR that does this (before passing it to MWDT strip). I am still a little worried that the Zephyr SDK strip isn't stripping correctly, since it's not designed to handle objects produced by MWDT CCAC, but I thought I would offer it as an alternative to people who don't like my script running for MWDT. The tests pass for nsim/nsim_vpx5, so... it works, at least.

laurenmurphyx64 committed:
llext: compile llext with arc mwdt on nsim vpx5

Allow users to compile LLEXTs with the MetaWare Design Toolchain.
Provides utility to dramatically shrink bloated extensions by removing
unused section names left behind by MWDT strip. Test extensions will
not fit in NSIM VPX5 memory otherwise; each is reduced from ~16k to ~1k.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 committed:
llext: run on harvard arch with static heaps

Splits LLEXT static heap into an instruction heap placed in
instruction memory and a data heap placed in data memory.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 committed:
llext: add reloc and symbols for mwdt

Adds relocation and symbols generated by MWDT while using it
for NSIM VPX5.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 committed:
llext: dynamic heap support for harvard arch

Adds support for LLEXT_HEAP_DYNAMIC with Harvard architecture.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 commented:
@M-Moawad I saw you submit an issue using ARC MWDT and LLEXT. If you like, you can take a brief look at these changes to make sure I don't break something you're using

pillo79 commented on subsys/llext/llext_mem.c:
```diff
@@ -28,10 +28,22 @@ LOG_MODULE_DECLARE(llext, CONFIG_LLEXT_LOG_LEVEL);
+#ifdef CONFIG_HARVARD
+struct k_heap llext_instr_heap;
+struct k_heap llext_data_heap;
+#else
 struct k_heap llext_heap;
+#endif
```
In the context of always having "separate" data/instr heaps for the different functions, even though they may be used as a single.
Wild idea, would it be possible to do this?
```suggestion
#else
struct k_heap llext_heap;
#define llext_instr_heap llext_heap
#define llext_data_heap llext_heap
#endif
```

would this reduce the #ifdefs in the rest of the code?

laurenmurphyx64 replied:
I went with Tom's suggestion, but if you prefer your suggestion, we can go with it instead since Tom said he doesn't have a strong preference

laurenmurphyx64 replied:
Made the change! Yeah, it looks a lot better

pillo79 commented on subsys/llext/llext_mem.c:
```diff
@@ -115,8 +127,22 @@ static int llext_copy_region(struct llext_loader *ldr, struct llext *ext,
-				if (IS_ALIGNED(ext->mem[mem_idx], region_align) ||
-				    ldr_parm->pre_located) {
+#ifdef CONFIG_HARVARD
+#ifdef CONFIG_ARC
+#define BASE_ADDR ext->mem[mem_idx]
+#define ALLOC region_alloc
+#else
+/* Check to be updated if any non-ARC boards using Harvard architecture is added;
+ * in the meantime, unconditionally copy text to instruction memory heap
+ */
+#define INSTR_FETCHABLE false
+#endif
```
INSTR_FETCHABLE is not defined in all cases in this #ifdef. Also the same code is in `llext_load.c`, it's best moved to a common header.

laurenmurphyx64 replied:
I have a common header in llext_priv.h, which defines a default true for INSTR_FETCHABLE - did I do it wrong?

laurenmurphyx64 replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/93627#discussion_r2240590364

pillo79 reviewed (commented):
Lots of good things in this huge PR, thanks! :+1: 

A few really early comments from my first reading:

* I think the reorder script is quite useful, and I'd like to see that as a separate PR, since it does apply to any LLEXT target. I always thought it would be possible to achieve that via a suitable LLEXT linker script (maybe target-specific), but never really looked into it. Is that not feasible?
Biggest downside I see with the linker script is that it needs to reuse lots of logic from the C side, so it may easily diverge.

* The more I think about the data/instr separation, the more I like it for _all_ targets. I think we should standardize `alloc_data` and `alloc_instr` functions (or maybe have an `area` argument?) Pluses:
  - it's a step towards clearly documenting the memory layout / usage in code, as it was recently discussed for #93274 ;
  - removes a lot of `#ifdef`s
  - we actually have a use case where we would like to _only_ allocate data; this would make it easier to add safety checks for that.

teburd commented:
I looked this over a few times today.

Truly impressive work. A lot of things were need to get here, I have *no* doubt. Like @pillo79 I think I'd like to see a little less ifdefry around the alloc situation. That's about my only nit and it is a nit/code readability thing less a structural thing.

e.g. perhaps as an alternative to the name aliasing with macros (I have no strong preference)

```c
void *llext_alloc_instr(size_t sz, size_t alignment)
{
#ifdef CONFIG_HARVARD
     struct k_heap *h = &llext_instr_heap;
#else
     struct k_heap *h = &llext_heap;
#endif

     return k_heap_alloc_aligned(h, sz, ...); 
 }
 ```

lyakh commented on subsys/llext/llext_mem.c:
```diff
@@ -254,7 +276,7 @@ int llext_copy_regions(struct llext_loader *ldr, struct llext *ext,
 			if (mem_idx < LLEXT_MEM_EXPORT) {
 				LOG_DBG("-s %s %#zx", name,
-					(size_t)ext->mem[mem_idx] + ldr->sect_map[i].offset);
+				(size_t)ext->mem[mem_idx] + ldr->sect_map[i].offset);
```
indentation

laurenmurphyx64 replied:
Fixed

lyakh commented on subsys/llext/llext_mem.c:
```diff
@@ -371,13 +376,23 @@ int llext_add_domain(struct llext *ext, struct k_mem_domain *domain)
+int llext_heap_init(void *instr_mem, size_t instr_bytes,
+		void *data_mem, size_t data_bytes)
+#else
 int llext_heap_init(void *mem, size_t bytes)
```
can we keep a single prototype? Just add the missing arguments to `llext_heap_init()` for everybody (LLEXT is still experimental, isn't it?) or if we absolutely want to keep the API, add another function and call it from the current `llext_heap_init()`?

laurenmurphyx64 replied:
I have updated it, but I don't know if I 100% understood what you suggested - let me know if it's right

laurenmurphyx64 committed:
llext: add fixes for various tests on mwdt / harvard arch

Adds fixes to get LLEXT tests to work for NSIM VPX on MWDT.

Excludes prelocated extensions for MWDT for now.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 committed:
llext: replace stripping script with call to zephyr sdk strip

Replace custom stripping script with a call to Zephyr SDK strip.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 commented on subsys/llext/llext_load.c:
```diff
@@ -312,6 +312,29 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext,
+/* Check to be updated if any non-ARC boards using Harvard architecture is added;
+ * in the meantime, proceed with a warning
+ */
+#define INSTR_FETCHABLE false
```
At the top of subsys/llext/llext_priv.h, which both llext_mem.c and llext_load.c include, I defined it to be true unless otherwise overridden - am I missing something?

laurenmurphyx64 commented on scripts/build/llext_elf_editor.py:
```diff
@@ -0,0 +1,57 @@
+    # ent_idx is 0 for ELF header
+    offset = (ent_idx * header.sizeof()) + get_field_offset(header, name)
+
+    field = None
```
Removed

laurenmurphyx64 commented on arch/arc/core/elf.c:
```diff
@@ -68,12 +69,28 @@ int arch_elf_relocate(struct llext_loader *ldr, struct llext *ext, elf_rela_t *r
 		break;
-	case R_ARC_S25W_PCREL:
-		/* ((S + A) - P) >> 2
+case R_ARC_S25H_PCREL:
```
Fixed

laurenmurphyx64 commented:
Along with other changes, I've pulled out the reordering script and will submit it as a separate PR.

laurenmurphyx64 commented on cmake/modules/extensions.cmake:
```diff
@@ -5965,9 +5965,24 @@ function(add_llext_target target_name)
   # Remove sections that are unused by the llext loader
   add_custom_command(
     OUTPUT ${llext_pkg_output}
+    COMMAND ${gnu_strip_for_mwdt_cmd}
     COMMAND $<TARGET_PROPERTY:bintools,elfconvert_command>
```
MWDT stripac has no support for section_remove (see EOF cmake/bintools/arcmwdt/elfconvert_command.cmake). Passing -qlu to stripac is what gets it to remove the 700 .arcextmap sections, leaving their 700 names in .shstrtab. Unfortunately, GNU strip is unable to detect these now hidden, unused names, so when I invoke it after stripac is done, it fails to remove them. So I unfortunately can't call stripac first (as far as I know from experimenting)

laurenmurphyx64 commented on include/zephyr/llext/llext.h:
```diff
@@ -412,6 +412,21 @@ int llext_get_section_header(struct llext_loader *loader, struct llext *ext,
+ * @param data_bytes Size of data memory region, in bytes
+ *
+ * @returns 0 on success, or a negative error code.
+ * @retval -ENOSYS Option @kconfig{CONFIG_LLEXT_HEAP_DYNAMIC} is not enabled or supported
```
@lyakh mentioned [here](https://github.com/zephyrproject-rtos/zephyr/pull/93627#discussion_r2235012824) to call one function from the other if I wanted to keep both, so technically, the Harvard one does work without Harvard arch, as it gets called by the original. Not sure if I 100% got the comment though

pillo79 replied:
Sorry, missed lyakh's comment when I wrote mine. 

Agree with not changing the parameter count on `llext_heap_init()`, but I don't see the advantage of calling `_init_harvard()`  from `_init()`: AFAICS on Harvard arches the current `_init()` will not work properly anyway.

If you really need extra code to support both, my preference would have been to make the choice explicit by returning `-ENOSYS` from the "wrong" one - hence the comment. 
At least, warn (or even error?) when calling `_init_harvard()` with two separate buffers on non-Harvard arches, since that will just waste memory.

 @lyakh what do you think?

laurenmurphyx64 commented on subsys/llext/llext_mem.c:
```diff
@@ -343,20 +348,34 @@ int llext_add_domain(struct llext *ext, struct k_mem_domain *domain)
+#ifdef CONFIG_LLEXT_HEAP_DYNAMIC
+	llext_heap_init_harvard(mem, bytes, NULL, 0);
+#else
+	return -ENOSYS;
```
Ugh! My bad

pillo79 reviewed (commented):
LGTM! Several tiny non-blocking nits, only two open questions: 
- `init`/`init_harvard` details?
- similarly, would it make sense to have a more uniform memory API? E.g. instead of having `heap_alloc` and `heap_alloc_instr`, have `heap_alloc_data`/`heap_alloc_instr` everywhere?

teburd commented on tests/subsys/llext/CMakeLists.txt:
```diff
@@ -74,7 +74,7 @@ get_property(COMPILER_TOPT TARGET compiler PROPERTY linker_script)
 set_ifndef(  TOPT -Wl,-T) # Use this if the compiler driver doesn't set a value
 
-if (CONFIG_LLEXT_TYPE_ELF_RELOCATABLE AND NOT CONFIG_ARM AND NOT CONFIG_RISCV)
+if (CONFIG_LLEXT_TYPE_ELF_RELOCATABLE AND NOT CONFIG_ARM AND NOT CONFIG_RISCV AND NOT COMPILER STREQUAL arcmwdt)
```
agree

laurenmurphyx64 commented on tests/subsys/llext/src/test_llext.c:
```diff
@@ -375,10 +376,22 @@ ZTEST(llext, test_inspect)
 	do_inspect_checks(ldr, ext, LLEXT_MEM_RODATA, ".rodata", "number_in_rodata");
 	do_inspect_checks(ldr, ext, LLEXT_MEM_RODATA, ".my_rodata", "number_in_my_rodata");
+#endif
 	do_inspect_checks(ldr, ext, LLEXT_MEM_TEXT, ".text", "function_in_text");
```
Changed

lyakh commented on subsys/llext/llext_mem.c:
```diff
@@ -343,13 +348,36 @@ int llext_add_domain(struct llext *ext, struct k_mem_domain *domain)
+	return 0;
+#else
+	return -ENOSYS;
+#endif
```
```
#if !defined(CONFIG_LLEXT_HEAP_DYNAMIC) || !defined(CONFIG_HARVARD)
	return -ENOSYS;
#else
	if (llext_heap_inited) {
		return -EEXIST;
	}

	k_heap_init(&llext_instr_heap, instr_mem, instr_bytes);
	k_heap_init(&llext_data_heap, data_mem, data_bytes);

	llext_heap_inited = true;
	return 0;
#endif
```
seems easier to me even if it changes the resulting error code in a (still invalid) call sequence
```
llext_heap_init();
llext_heap_init_harvard();
```

laurenmurphyx64 replied:
Changed! Looks a lot cleaner, thanks

lyakh commented on subsys/llext/llext_mem.c:
```diff
@@ -343,13 +348,36 @@ int llext_add_domain(struct llext *ext, struct k_mem_domain *domain)
 	if (llext_heap_inited) {
 		return -EEXIST;
 	}
+#ifdef CONFIG_HARVARD
```
similar to the above
